File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/transform_matcher.rb
require 'cucumber/cucumber_expressions/transform_matcher'

module Cucumber
  module CucumberExpressions
    class TransformMatcher
      attr_reader :transform

      def initialize(transform, regexp, text, match_position=0)
        @transform, @regexp, @text = transform, regexp, text
        @match = @regexp.match(@text, match_position)
      end

      def advance_to(new_match_position)
        self.class.new(@transform, @regexp, @text, new_match_position)
      end

      def find
        !@match.nil?
      end

      def start
        @match.begin(0)
      end

      def group
        @match.captures[0]
      end

      def <=>(other)
        pos_comparison = start <=> other.start
        return pos_comparison if pos_comparison != 0
        length_comparison = other.group.length <=> group.length
        return length_comparison if length_comparison != 0
        0
      end
    end
  end
end

File: cucumber-expressions/javascript/src/transform_matcher.js
class TransformMatcher {

  constructor(transform, captureGroupRegexp, text, matchPosition) {
    this._transform = transform
    this._captureGroupRegexp = captureGroupRegexp
    this._text = text
    this._matchPosition = matchPosition || 0

    const regexp = new RegExp(`(${captureGroupRegexp})`)
    this._match = regexp.exec(text.slice(this._matchPosition))
  }

  get transform() {
    return this._transform
  }

  advanceTo(newMatchPosition) {
    return new TransformMatcher(this._transform, this._captureGroupRegexp, this._text, newMatchPosition)
  }

  get find() {
    return this._match
  }

  get start() {
    return this._matchPosition + this._match.index
  }

  get group() {
    return this._match[1]
  }

  static compare(a, b) {
    const posComparison = a.start - b.start
    if (posComparison != 0) return posComparison
    const lengthComparison = b.group.length - a.group.length
    if (lengthComparison != 0) return lengthComparison
    return 0
  }
}

export default TransformMatcher
